File: ios/Classes/AdyenFlutterPlugin.swift
import Flutter
import UIKit
import AdyenPOS

public class SwiftAdyenFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, ConnectionObserver {

    private var discoveredReaders: [String: Reader] = [:]
    private var eventSink: FlutterEventSink?
    


    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "adyen_flutter", binaryMessenger: registrar.messenger())
        let instance = SwiftAdyenFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
       
        let eventChannel = FlutterEventChannel(name: "adyen_flutter/events", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)

    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }


    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initializeSdk":
            if let args = call.arguments as? [String: Any],
               let environmentStr = args["environment"] as? String,
               let merchantAccount = args["merchantAccount"] as? String,
               let clientKey = args["clientKey"] as? String {
                initializeSdk(environmentStr: environmentStr, merchantAccount: merchantAccount, clientKey: clientKey)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing args", details: nil))
            }
        case "discoverReaders":
            discoverReaders(result: result)

        case "connectToReader":
            guard let args = call.arguments as? [String: Any],
                  let serialNumber = args["serialNumber"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing serialNumber", details: nil))
                return
            }
            connectToReader(serialNumber: serialNumber, result: result)

        case "startPayment":
            guard let args = call.arguments as? [String: Any],
                  let amount = args["amount"] as? Int,
                  let currency = args["currency"] as? String,
                  let reference = args["reference"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing payment args", details: nil))
                return
            }
            startPayment(amount: amount, currency: currency, reference: reference, result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initializeSdk(environmentStr: String, merchantAccount: String, clientKey: String) {
        let environment: Environment = environmentStr.uppercased() == "LIVE" ? .live : .test

        let configuration = TerminalConfiguration(
            environment: environment,
            clientKey: clientKey,
            merchantAccount: merchantAccount
        )

        Terminal.initialize(configuration: configuration)
    }

    private func discoverReaders(result: @escaping FlutterResult) {
        let discoveryConfiguration = ReaderDiscoveryConfiguration(readerTypes: [.any], scanMethod: .bluetooth)

        Terminal.shared.discoverReaders(discoveryConfiguration) { readers, error in
            if let error = error {
                result(FlutterError(code: "DISCOVERY_ERROR", message: error.localizedDescription, details: nil))
                return
            }

            let readerList = readers.map { reader in
                return [
                    "name": reader.model,
                    "serialNumber": reader.serialNumber
                ]
            }

            result(readerList)
        }
    }


    private func connectToReader(serialNumber: String, result: @escaping FlutterResult) {
        guard let reader = discoveredReaders[serialNumber] else {
            result(FlutterError(code: "READER_NOT_FOUND", message: "Reader not found: \(serialNumber)", details: nil))
            return
        }

        Terminal.shared.connectReader(reader) { connectedReader, error in
            if let error = error {
                self.eventSink?("connection_failed: \(error.localizedDescription)")
                result(FlutterError(code: "CONNECTION_FAILED", message: error.localizedDescription, details: nil))
                return
            }

            guard let _ = connectedReader else {
                self.eventSink?("connection_failed: unknown error")
                result(FlutterError(code: "CONNECTION_FAILED", message: "Unknown error", details: nil))
                return
            }

            Terminal.shared.addConnectionObserver(self)

            self.eventSink?("connected")
            result(true)
        }
    }

    public func didConnect(reader: Reader) {
        eventSink?("connected")
    }

    public func didDisconnect(reader: Reader, error: Error?) {
        if let error = error {
            eventSink?("disconnected: \(error.localizedDescription)")
        } else {
            eventSink?("disconnected")
        }
    }
                    
    private func startPayment(amount: Int, currency: String, reference: String, result: @escaping FlutterResult) {
        let paymentAmount = Amount(currency: currency, value: amount)
        let paymentRequest = PaymentRequest(amount: paymentAmount, reference: reference)

        Terminal.shared.submitPayment(paymentRequest) { paymentResult, error in
            if let error = error {
                result(FlutterError(code: "PAYMENT_FAILED", message: error.localizedDescription, details: nil))
                return
            }

            guard let paymentResult = paymentResult else {
                result(FlutterError(code: "PAYMENT_FAILED", message: "Unknown error", details: nil))
                return
            }

            var status: String

            switch paymentResult.result {
            case .authorized:
                status = "captured"
            case .declined:
                status = "declined"
            case .cancelled:
                status = "cancelled"
            @unknown default:
                status = "unknown"
            }

            let paymentData: [String: Any] = [
                "status": status,
                "pspReference": paymentResult.pspReference
            ]

            result(paymentData)
        }
    }

    deinit {
        Terminal.shared.removeConnectionObserver(self)
    }
}

File: android/src/main/kotlin/com/example/adyen_flutter/AdyenFlutterPlugin.kt
package com.yourcompany.adyen_flutter

import android.app.Application
import androidx.annotation.NonNull
import com.adyen.checkout.terminal.Terminal
import com.adyen.checkout.terminal.TerminalConfiguration
import com.adyen.checkout.terminal.Environment

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class AdyenFlutterPlugin: FlutterPlugin, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var terminal: Terminal
    private val discoveredReaders = mutableMapOf<String, Reader>()

    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "adyen_flutter")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "adyen_flutter/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        when (call.method) {
            "initializeSdk" -> {
                val environmentStr = call.argument<String>("environment") ?: "TEST"
                val merchantAccount = call.argument<String>("merchantAccount") ?: ""
                val clientKey = call.argument<String>("clientKey") ?: ""

                initializeSdk(environmentStr, merchantAccount, clientKey)
                result.success(null)
            }

            "discoverReaders" -> {
                discoverReaders(result)
            }

            "connectToReader" -> {
                val serialNumber = call.argument<String>("serialNumber") ?: ""
                connectToReader(serialNumber, result)
            }

            "startPayment" -> {
                val amount = call.argument<Int>("amount") ?: 0
                val currency = call.argument<String>("currency") ?: "USD"
                val reference = call.argument<String>("reference") ?: "flutter-txn"

                startPayment(amount, currency, reference, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initializeSdk(environmentStr: String, merchantAccount: String, clientKey: String) {
        val env = if (environmentStr == "LIVE") Environment.LIVE else Environment.TEST

        val terminalConfig = TerminalConfiguration.Builder(
            environment = env,
            clientKey = clientKey,
            merchantAccount = merchantAccount
        ).build()

        Terminal.init(terminalConfig)
    }

    private fun discoverReaders(result: MethodChannel.Result) {
        val configuration = ReaderDiscoveryConfiguration(ReaderType.ALL, ScanMethod.BLUETOOTH)

        terminal.discoverReaders(configuration, object : ReaderDiscoveryListener {
            override fun onReaderDiscovered(reader: Reader) {
                val readerInfo = mapOf(
                    "name" to reader.model,
                    "serialNumber" to reader.serialNumber
                )
                // Send list of readers (or use EventChannel for real-time discovery)
                result.success(listOf(readerInfo))  // Simplified. Ideally use EventChannel or cache readers.
            }

            override fun onDiscoveryStopped() {
                // Optional: Notify discovery stopped.
            }
        })
    }

    private fun connectToReader(serialNumber: String, result: MethodChannel.Result) {
        val reader = discoveredReaders[serialNumber]

        if (reader == null) {
            result.error("READER_NOT_FOUND", "Reader not found with serial: $serialNumber", null)
            return
        }

        terminal.connectReader(reader, object : ConnectionListener {
            override fun onConnect(reader: Reader) {
                result.success(true) 
                eventSink?.success("connected") 
            }

            override fun onDisconnect(reader: Reader) {
                eventSink?.success("disconnected") 
            }

            override fun onFailure(reader: Reader, error: TerminalException) {
                result.error("CONNECTION_FAILED", error.message, null)
                eventSink?.success("connection_failed: ${error.message}") 
            }
        })
    }


    private fun startPayment(amount: Int, currency: String, reference: String, result: MethodChannel.Result) {
        val paymentAmount = Amount.Builder()
            .currency(currency)
            .value(amount.toLong())  // Adyen expects long for minor units
            .build()

        val paymentRequest = PaymentRequest.Builder(paymentAmount, reference).build()

        terminal.submitPayment(paymentRequest, object : PaymentResultListener {
            override fun onPaymentCompleted(paymentResult: PaymentResult) {
                val status = when (paymentResult.result) {
                    PaymentResult.Result.AUTHORIZED -> "captured"
                    PaymentResult.Result.DECLINED -> "declined"
                    PaymentResult.Result.CANCELLED -> "cancelled"
                    else -> "unknown"
                }

                val paymentData = mapOf(
                    "status" to status,
                    "pspReference" to paymentResult.pspReference
                )

                result.success(paymentData)
            }


            override fun onPaymentFailed(error: TerminalException) {
                result.error("PAYMENT_FAILED", error.message, null)
            }
        })
    }


    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
